README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# agvgd <img src='man/figures/logo.svg' align="right" height="139" />

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/agvgd)](https://CRAN.R-project.org/package=agvgd)
[![Codecov test
coverage](https://codecov.io/gh/maialab/agvgd/branch/master/graph/badge.svg)](https://app.codecov.io/gh/maialab/agvgd?branch=master)
<!-- badges: end -->

The R package `{agvgd}` provides an R implementation of the
Align-GVGD<sup>[1](#1),[2](#2),[3](#3)</sup> (A-GVGD) method.

A-GVGD combines multiple sequence alignment of orthologous sequences
with the Grantham distance<sup>[4](#4)</sup> to classify missense
variants, i.e. to distinguish human disease susceptibility missense
changes from changes of little clinical significance.

## Installation

Install `{agvgd}` from CRAN:

``` r
install.packages("agvgd")
```

You can install the development version of `{agvgd}` like so:

``` r
# install.packages("remotes")
remotes::install_github("maialab/agvgd")
```

## Usage

Here is a minimal example using a dummy alignment constructed from an R
matrix:

``` r
library(agvgd)

# The alignment can be provided as a simple matrix:
alignment <- matrix(
  c('P', 'M', 'I',
    'P', 'I', 'I',
    'P', 'L', 'I'),
  nrow = 3,
  byrow = TRUE
)

# In this example we will interrogate changes to the second position in the
# alignment, i.e., changes to Methionine (M); therefore we set the position of
# interest (poi) to 2.
poi <- 2

# Here's a set of three possible missense changes (substitutions): Isoleucine
# (Ile, I), Leucine (Leu, L), and Tryptophan (Trp, W):
substitutions <- c('I', 'L', 'W')

# agvgd package's main function is `agvgd()` :)
agvgd(alignment = alignment, poi = poi, sub = substitutions)
#> # A tibble: 3 × 7
#>     res   poi ref   sub      gv    gd prediction
#>   <int> <int> <chr> <chr> <dbl> <dbl> <chr>     
#> 1     2     2 M     I      14.3   0   C0        
#> 2     2     2 M     L      14.3   0   C0        
#> 3     2     2 M     W      14.3  60.4 C35
```

This is another simple example but using one of the bundled alignments
with `{agvgd}`. This one is the alignment for the gene ATM. Let’s say
you are interested in position 16 of the alignment, and that you would
like to know the impact of all possible missense substitutions:

``` r
# `read_alignment()` either reads in one of the bundled alignments with
# `{agvgd}` by gene name or directly from a FASTA file.
alignment <- read_alignment(gene = 'ATM')

# For the sake of illustration, let us define a new alignment based on a narrow
# region of the original alignment by focusing only on the first 30 positions.
# Note: rows are protein sequences; columns are alignment positions.
narrow_alignment <- alignment[, 1:30]

# poi: position of interest
poi <- 16

# Print the alignment to the console and highlight the position of interest
# (POI). NB: The POI is not highlighted in GitHub's README but it is on the R
# console.
print(narrow_alignment, poi = poi)
#> Hsap_ATM_AAB65827.1  1 M-SLVLNDLLICCRQLEHDRATERKKEVEK
#> Mmus_ATM_NP_031525.2 1 M-SLALNDLLICCRQLEHDRATERRKEVDK
#> Sscr_ATM_AAT01608.1  1 M-SLALNDLLICCRQLEHDRATERRKAVEN
#> Mdom_ATM_IARC        1 M-SLALNDLLLCCRQLENDRATERRKEVEK
#> Ggal_ATM_edited      1 M-SLVLHDLLTCCRRLENERATERRNEIEN
#> Xlae_ATM_AAT72929.1  1 M-SLALHELLLCCRQIETDKATERKKEIVK
#> Drer_ATM_IARC_v2     1 M-SLALHELLVCCRGLENEKATERKKEVDR
#> Bflo_ATM_IARC        1 MTDLLTHDLRDCCCHLESDKVTERKKNAEK
#> Spur_ATM_ABY60856.1  1 MAEVLIP-LRTACGYLGSDKITERKKQIDI

# You may use `amino_acids()` to get a vector of all the 20 standard amino acids
# and hence easily get a vector of all possible substitutions:
all_substitutions <- amino_acids()

agvgd(alignment = narrow_alignment, poi = poi, sub = all_substitutions)
#> # A tibble: 20 × 7
#>      res   poi ref   sub      gv    gd prediction
#>    <int> <int> <chr> <chr> <dbl> <dbl> <chr>     
#>  1    15    16 L     S      4.86 142.  C65       
#>  2    15    16 L     R      4.86  97.6 C65       
#>  3    15    16 L     L      4.86   0   C0        
#>  4    15    16 L     P      4.86  95.4 C65       
#>  5    15    16 L     T      4.86  89.3 C65       
#>  6    15    16 L     A      4.86  93.7 C65       
#>  7    15    16 L     V      4.86  29.6 C15       
#>  8    15    16 L     G      4.86 135.  C65       
#>  9    15    16 L     I      4.86   0   C0        
#> 10    15    16 L     F      4.86  21.3 C0        
#> 11    15    16 L     Y      4.86  33.0 C25       
#> 12    15    16 L     C      4.86 197.  C65       
#> 13    15    16 L     H      4.86  94.3 C65       
#> 14    15    16 L     Q      4.86 109.  C65       
#> 15    15    16 L     N      4.86 149.  C65       
#> 16    15    16 L     K      4.86 102.  C65       
#> 17    15    16 L     D      4.86 168.  C65       
#> 18    15    16 L     E      4.86 134.  C65       
#> 19    15    16 L     M      4.86  10.1 C0        
#> 20    15    16 L     W      4.86  60.5 C55
```

## Disclaimer

The data and information contained herein and in the results of the
`{agvgd}` package are provided on an *as is* basis and the authors makes
no representations or warranties, either expressed or implied, as to
their accuracy, completeness or suitability for a particular purpose.
Similarly, authors make no representations or warranties with regard to
the non-infringement of third party proprietary rights. Thus, the
authors do not accept any responsibility or liability with regard to the
reliance on, and/or use of, such data and information.

## Logo

The `{agvgd}` logo, `agvgd.png`, is a derivative work of an illustration
of [“Globin Evolution”](https://pdb101.rcsb.org/motm/206) by David S.
Goodsell and the [RCSB PDB](https://www.rcsb.org/), used under
[CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/). `agvgd.png`
is licensed under CC-BY-4.0 by Ramiro Magno.

## Related software

-   Align-GVGD (A-GVGD) by [Sean
    Tavtigian](https://uofuhealth.utah.edu/huntsman/labs/tavtigian/),
    provided as a web service and hosted by the [Huntsman Cancer
    Institute](https://healthcare.utah.edu/huntsmancancerinstitute/) at
    <http://agvgd.hci.utah.edu/>.
-   Multivariate Analysis of Protein Polymorphism
    (MAPP)<sup>[5](#5)</sup> by [Eric
    Stone](https://bdsi.anu.edu.au/people/professor-eric-stone) that had
    the same ideas as A-GVGD but in a more sophisticated framework. This
    used to be provided as a Java standalone application:
    <https://biology.anu.edu.au/research/groups/stone-group-quantitative-and-computational-biology/software>;
    however, the download link to MAPP has been down for a long while.
-   The [grantham](https://cran.r-project.org/package=grantham) package.

## Acknowledgements

The work on this package benefited greatly from the feedback of
Professor Sean Tavtigian, and Dr. Russell Bell.

## References

<a id="1">1.</a> Tavtigian, S.V., Deffenbaugh, A. M., Yin, L., Judkins,
T., Scholl, T., Samollow, P.B., de Silva, D., Zharkikh, A., Thomas, A.
*Comprehensive statistical study of 452 BRCA1 missense substitutions
with classification of eight recurrent substitutions as neutral*.
Journal of Medical Genetics 43, 295–305 (2006). doi:
[10.1136/jmg.2005.033878](https://doi.org/10.1136/jmg.2005.033878).

<a id="2">2.</a> Mathe, E., Olivier, M., Kato, S., Ishioka, C., Hainaut,
P., Tavtigian, S.V. *Computational approaches for predicting the
biological effect of p53 missense mutations: a comparison of three
sequence analysis based methods*. Nucleic Acids Research 34, 1317–1325
(2006). doi: [10.1093/nar/gkj518](https://doi.org/10.1093/nar/gkj518).

<a id="3">3.</a> Tavtigian, S.V., Byrnes, G. B, Goldgar, D. E., Thomas,
A. *Classification of rare missense substitutions, using risk surfaces,
with genetic- and molecular-epidemiology applications*. Human Mutation
29, 1342–1354. doi:
[10.1002/humu.20896](https://doi.org/10.1002/humu.20896)

<a id="3">3.</a> Grantham, R. *Amino acid difference formula to help
explain protein evolution*. Science 185, 862–864 (1974). doi:
[10.1126/science.185.4154.862](https://doi.org/10.1126/science.185.4154.862).

<a id="4">4.</a> Stone, E. A., Sidow, Arend. *Physicochemical constraint
violation by missense substitutions mediates impairment of protein
function and disease severity*. Genome Research 15, 978–986 (2005). doi:
[10.1101/gr.3804205](https://doi.org/10.1101/gr.3804205).


NEWS.md
# agvgd 0.1.2

* Fixed a bug with `read_substitutions()` that would not parse positions correctly.
* Added a new substitutions example file `lee2010_sub.txt`.
* Added a new function `read_agvgdweb_results()` aimed at developers.

# agvgd 0.1.1

* The main function `agvgd()` had an alias, `align_agvgd()`, which is now removed to avoid confusion.
* A few more functions are now exported allowing access to low-level routines: `gd()`, `gv()` and `cpv_ranges()`.
* `alignment_pos_to_residue_pos()` has been renamed to `poi_to_res()` and `residue_pos_to_alignment_pos()` to `res_to_poi()`.
* Code coverage is now tracked.
* Added a `NEWS.md` file to track changes to the package.

# agvgd 0.1.0

* First version of `{agvgd}`.


R/agvgd.R
#' Align-GVGD (A-GVGD)
#'
#' @description
#' This function implements the Align-GVGD (A-GVGD) method described in
#' Tavtigian _et al._ (2006).
#'
#' A-GVGD combines multiple sequence alignment of orthologous sequences with the
#' Grantham distance to classify missense variants, i.e. to distinguish human
#' disease susceptibility missense changes from changes of little clinical
#' significance.
#'
#' The biochemical variation at each alignment position is converted to a
#' Grantham Variation score (GV) and the difference between these properties and
#' those of the variant amino acid being assessed are calculated and a Grantham
#' Difference score generated (GD).  The predicted effect is classed as C0, C15,
#' C25, C35, C45, C55, or C65, with C65 most likely to interfere with function
#' and C0 least likely.
#'
#' @param alignment A character matrix or an alignment object obtained with
#'   [read_alignment()]. Rows are expected to be sequences of single characters
#'   (protein residues), and columns the alignment positions. The first row must
#'   be the reference sequence, i.e. the sequence whose substitutions will be
#'   evaluated against.
#' @param poi A whole number indicating the position of interest (POI).
#' @param sub A character vector of protein residue substitutions to be
#'   classified. The amino acids must be provided as one-letter symbols.
#' @param mode If both `poi` and `sub` contain more than one element, `mode`
#'   specifies how these two inputs are combined. If `mode = 'recycle'` the
#'   shortest vector is recycled to match the length of the longest. If `mode =
#'   'expand_grid'`, all combinations between elements of `poi` and `sub` are
#'   combined.
#' @param sort Whether to sort the output by `gd`, or not. Default is `FALSE`.
#' @param keep_self Whether to keep those results in the output that correspond
#'   to residues being the same in `ref` and `sub`. Default is `TRUE`. But it
#'   will be useful to change it to `FALSE` if want to compare the results with
#'   those provided by <http://agvgd.hci.utah.edu/> that filters them out.
#' @param digits Integer indicating the number of decimal places to be used in
#'   rounding `gv` and `gd` values. Default is `2`. Note that the calculation of
#'   the `prediction` variable won't be affected by rounding of `gv` and `gd`,
#'   as it is calculated prior to the rounding.
#'
#' @return A [tibble][tibble::tibble-package] whose observations refer to the
#'   combination alignment position and amino acid substitution; consists of
#'   seven variables:
#' \describe{
#'   \item{res}{Position of the amino acid residue in the reference protein
#'   (first sequence in the alignment). This position corresponds to `poi` minus
#'   the gaps in the alignment.}
#'   \item{poi}{Position of interest, i.e. the alignment position at which the
#'   amino acid substitution is being assessed.}
#'   \item{ref}{Reference amino acid, i.e. the amino acid in the first sequence
#'   of the alignment, at the position of interest.}
#'   \item{sub}{Amino acid substitution being assessed.}
#'   \item{gv}{Grantham variation score.}
#'   \item{gd}{Grantham difference score.}
#' \item{prediction}{Predicted effect of the amino acid substitution. This is
#' classed as C0, C15, C25, C35, C45, C55, or C65, with C65 most likely to
#' interfere with function and C0 least likely.}
#' }
#'
#' @references
#' - Tavtigian, S.V., Deffenbaugh, A. M., Yin, L., Judkins, T., Scholl, T.,
#' Samollow, P.B., de Silva, D., Zharkikh, A., Thomas, A. _Comprehensive
#' statistical study of 452 BRCA1 missense substitutions with classification of
#' eight recurrent substitutions as neutral_. Journal of Medical Genetics 43,
#' 295--305 (2006). \doi{10.1136/jmg.2005.033878}.
#' - Mathe, E., Olivier, M., Kato, S., Ishioka, C., Hainaut, P., Tavtigian, S.V.
#' _Computational approaches for predicting the biological effect of p53
#' missense mutations: a comparison of three sequence analysis based methods_.
#' Nucleic Acids Research 34, 1317--1325 (2006). \doi{10.1093/nar/gkj518}.
#'
#' @md
#' @importFrom rlang .data
#'
#' @examples
#' # Read an alignment into R, e.g. the alignment for gene ATM.
#' alignment_ATM <- read_alignment(gene = 'ATM')
#'
#' # Predict the impact of changing the first residue (Met) to a Serine (S).
#' agvgd(alignment = alignment_ATM, poi = 1, sub = 'S')
#'
#' # `poi` can be a vector of positions, e.g., 3 thru 10, allow for prediction
#' # of multiple positions at once.
#' agvgd(alignment = alignment_ATM, poi = 3:10, sub = 'S')
#'
#' # `poi` expects a position in the frame of reference of the alignment, i.e.
#' # an alignment position (a column index). However, if you know instead
#' # the residue position in the reference sequence (first sequence in the
#' # alignment), then you may use the function `res_to_poi()`
#' # to convert from residue position to alignment position.
#' #
#' # Example: The second residue in the reference sequence of the ATM alignment
#' # is a Serine, after a Methionine. In the alignment, there is a gap between
#' # the two residues, so the alignment is 3 but the residue position on the
#' # protein is 2.
#' (poi2 <- res_to_poi(alignment_ATM, 2))
#' agvgd(alignment = alignment_ATM, poi = poi2, sub = 'A')
#'
#' # Because changes are context-dependent, i.e. they depend on the residue
#' # variation observed at a given alignment position, the same reference
#' # residue when replaced with the same substitution will in general have
#' # a different predicted impact.
#' agvgd(alignment = alignment_ATM, poi = 9:10, sub = 'S')
#'
#' # Use the ancillary function `amino_acids()` to get a vector of one-letter
#' # residue substitutions if you want to quickly assess the impact of all
#' # possible substitutions.
#' agvgd(alignment = alignment_ATM, poi = 1, sub = amino_acids())
#'
#' # Parameter `mode` gives you flexibility on how to combine `poi` and `sub`.
#' agvgd(alignment = alignment_ATM, poi = 3:4, sub = c('A', 'V'))
#'
#' # Use 'expand_grid' for all combinations.
#' agvgd(alignment = alignment_ATM, poi = 3:4, sub = c('A', 'V'), mode = 'expand_grid')
#'
#' @export
agvgd <- function(alignment,
                  poi,
                  sub,
                  mode = c('recycle', 'expand_grid'),
                  sort = FALSE,
                  keep_self = TRUE,
                  digits = 2L) {

  mode <- match.arg(mode)

  if(!rlang::is_integerish(poi))
    stop('`poi` can\'t be interpreted as a set of alignment positions.')

  if(!are_all_one_letter(sub))
    stop('`sub` must be a character vector of one-letter amino acid codes.')

  # We allow `poi` to be a double if it is integerish, here we convert it
  # to integer to make it formally an integer in case it was a double.
  poi <- as.integer(poi)

  if (identical(mode, 'recycle')) {
    n <- max(length(poi), length(sub))
    poi <- rep(poi, length.out = n)
    sub <- rep(sub, length.out = n)
    tbl <- tidyr::tibble(poi = poi, sub = sub)

  } else { # expand_grid (cartesian product)
    tbl <- tidyr::expand_grid(poi, sub)
  }

  agvgd_output <- purrr::map2_dfr(
    .x = tbl$poi,
    .y = tbl$sub,
    .f = ~ agvgd_(
      alignment = alignment,
      poi = .x,
      sub = .y
    )
  )

  # Rounding of `gv` and `gd`
  agvgd_output <-
    dplyr::mutate(agvgd_output,
                  gv = round(.data$gv, digits = digits),
                  gd = round(.data$gd, digits = digits))

  if(sort) {
    agvgd_output <- dplyr::arrange(agvgd_output, dplyr::desc(.data$gd))
  }

  if (!keep_self) {
    agvgd_output <- dplyr::filter(agvgd_output, .data$ref != .data$sub)
  }

  return(agvgd_output)
}

# `agvgd_()` is where the main logic of Align-GVGD is laid out.
agvgd_ <-
  function(alignment,
           poi,
           sub) {

  check_alignment(alignment)
  check_poi(poi)
  check_substitution(sub)

  alignment_at_poi <- filter_alignment(alignment, poi)

  # Find ranges for `c`, `p` and `v`.
  r <- cpv_ranges(alignment_at_poi)

  # Update min and max values according to the number of gaps
  r <- cpv_ranges_after_gaps(r, alignment_at_poi)

  gv <- gv(
    c_min = r$c_min,
    c_max = r$c_max,
    p_min = r$p_min,
    p_max = r$p_max,
    v_min = r$v_min,
    v_max = r$v_max
  )

  gd <- gd(
    sub,
    c_min = r$c_min,
    c_max = r$c_max,
    p_min = r$p_min,
    p_max = r$p_max,
    v_min = r$v_min,
    v_max = r$v_max
  )

  prediction <- gvgd_to_class(gv, gd)
  ref <- stats::setNames(alignment_at_poi[1, 2], NULL)

  tbl <- tibble::tibble(
    res = poi_to_res(
      alignment = alignment,
      poi = poi),
    poi = poi,
    ref = ref,
    sub = sub,
    gv,
    gd,
    prediction
  )

  return(tbl)

}

check_alignment <- function(alignment) {
  TRUE
}

check_poi <- function(poi) {
  TRUE
}

check_substitution <- function(substitution) {
  TRUE
}
